### PR TITLE
Adds mix_project_config[:xref][:exclude] for Xref

### PR DESCRIPTION
**Pull request by antipax**
#4661

/cc @josevalim

### CODE REVIEW

File: lib/mix/lib/mix/tasks/xref.ex [hunk 1]
```diff
@@ -24,6 +24,14 @@ defmodule Mix.Tasks.Xref do
     * `--no-archives-check` - do not check archives
     * `--no-elixir-version-check` - do not check the Elixir version from mix.exs
 
+  ## Configuration
+
+    All configuration for Xref should be placed under the key `:xref`.
+
+    * `:exclude` - modules and `{module, func, arity}`s to ignore when checking
+      cross references. For example:
+      `xref: [exclude: [MissingModule, {MissingModule2, :missing_func, 2}]]`
```
Comment (whatyouhide) [inline lib/mix/lib/mix/tasks/xref.ex]: I would not show the whole `:xref -> :exclude -> ...` example as we're describing just the `:exclude` key. I'd just show "For example `[MissingModule, {OtherMissingModule, :missing_func, 2}]`".

File: lib/mix/lib/mix/tasks/xref.ex [hunk 2]
```diff
@@ -24,6 +24,14 @@ defmodule Mix.Tasks.Xref do
     * `--no-archives-check` - do not check archives
     * `--no-elixir-version-check` - do not check the Elixir version from mix.exs
 
+  ## Configuration
+
+    All configuration for Xref should be placed under the key `:xref`.
+
+    * `:exclude` - modules and `{module, func, arity}`s to ignore when checking
```
Comment (whatyouhide) [inline lib/mix/lib/mix/tasks/xref.ex]: Maybe we can rephrase this to be

> a list of modules and `{module, function, arity}` tuples [...]

Wdyt?

File: lib/mix/lib/mix/tasks/xref.ex [hunk 3]
```diff
@@ -72,25 +80,29 @@ defmodule Mix.Tasks.Xref do
   ## Unreachable
 
   defp unreachable(pair) do
+    excludes = excludes()
+
     for manifest <- E.manifests(),
         {_, sources} = parse_manifest(manifest),
         source(source: file) = source <- sources,
-        entries = unreachable_source(source),
+        entries = unreachable_source(source, excludes),
```
Comment (whatyouhide) [inline lib/mix/lib/mix/tasks/xref.ex]: Do we need to have `excludes` in a variable, or can we just call `excludes/0` here?

Comment (josevalim) [inline lib/mix/lib/mix/tasks/xref.ex]: Variable is better so we don't read it over and over.

Comment (whatyouhide) [inline lib/mix/lib/mix/tasks/xref.ex]: Ops, missed that, good catch.

**Comment by josevalim** [comment]
This looks perfect. Let's just fix the docs as @whatyouhide suggested and rebase! (the need to rebase is probably my fault :D)

File: lib/mix/lib/mix/tasks/xref.ex [hunk 4]
```diff
@@ -75,22 +82,24 @@ defmodule Mix.Tasks.Xref do
     for manifest <- E.manifests(),
         {_, sources} = parse_manifest(manifest),
         source(source: file) = source <- sources,
-        entries = unreachable_source(source),
+        entries = unreachable_source(source, excludes()),
```
Comment (josevalim) [inline lib/mix/lib/mix/tasks/xref.ex]: The variable approach was better. :)

**Comment by antipax** [comment]
@whatyouhide is still a cool dude with good ideas but variable re-added. :)

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
